### DISCUSSION TITLE
Is the correct main bucket (and thus main bucket manifests) used if the user never invoked the `scoop bucket add main` command?

### DISCUSSION THREAD

**Discussion by Gitoffthelawn** [ScoopInstaller/Scoop#6172] [Help and Questions]
This question is related to issue #6171.

If the user never invokes a `scoop bucket add main` command, a main bucket is still used, but the URI it points to is different than after the user invokes the aforementioned command.

In this case (when the user has never invoked the `scoop bucket add main` command), is the correct main bucket (and thus the correct main bucket manifests) always used?

**Comment by HUMORCE** [comment]
[The installer](https://github.com/ScoopInstaller/Install) downloads ZIP file of Scoop/Main bucket repo. Then, when you run `scoop update|install|etc.`, Scoop tries to turn itself and the main bucket into a git repo, if `git` is installed.

- The main bucket that comes with fresh installation of Scoop is the latest commit of [main bucket repo](https://github.com/ScoopInstaller/Main).
- Scoop will not verify the remote url of bucket.
- In some cases, you may lose the main bucket on the first install. when that happens, you should fix it manually or just reinstall Scoop.

**Reply by Gitoffthelawn** [reply]
Thank you for your quick and helpful reply.  Follow-ups:

1. What you wrote about Scoop trying to turn itself into a git repo matches my recollection.  If I recall correctly, Scoop couldn't perform its most common uses until `git` was installed. This surprised me, but Scoop gave clear instructions on how to install `git`.  Why does it need to perform this conversion to a git repo?
2. You mentioned that the main bucket that comes with a fresh install of Scoop is the latest commit of the main bucket repo.  To me, that means it is the same main bucket (and thus the correct manifests) that would be added if the user invoked `scoop bucket add main`.  Is that understanding correct?
3. You mentioned that "Scoop will not verify the remote url of bucket".   Is that both on a fresh scoop install and after invoking `scoop bucket add main`?  Would it be beneficial to verify the URL, or would that serve no benefit?
4. Why would the system ever lose the main bucket on the first install? If it should happen, besides reinstalling Scoop, how would one fix it manually?

**Reply by HUMORCE** [reply]
> Why does it need to perform this conversion to a git repo?

Rolling updates of Scoop and buckets.

> You mentioned that the main bucket that comes with a fresh install of Scoop is the latest commit of the main bucket repo.

Yes, latest commit relative to time of Installation complete.

> that means it is the same main bucket (and thus the correct manifests) that would be added if the user invoked scoop bucket add main. Is that understanding correct?

Yes, for contents of bucket, the following 4 command lines are equivalent.
```
> scoop bucket add main
> scoop bucket add main https://github.com/ScoopInstaller/Main
> scoop bucket add main https://github.com/ScoopInstaller/Main.git
> scoop bucket add main git@github.com:ScoopInstaller/Main.git
```

> You mentioned that "Scoop will not verify the remote url of bucket". Is that both on a fresh scoop install and after invoking scoop bucket add main?

Since Scoop is installed on user-side as source code. Fresh installation and `scoop bucket add <well_known_bucket>` will both pull the correct buckets.

> Would it be beneficial to verify the URL, or would that serve no benefit?

As long as Scoop's code has not been modified(maliciously), verification is not necessary. MITM attacks are not considered, it's almost impossible.

> Why would the system ever lose the main bucket on the first install? If it should happen, besides reinstalling Scoop, how would one fix it manually?

Networking Issues.

Download it as ZIP from https://github.com/ScoopInstaller/Main, then extract inner folder `main-master` to `~\scoop\buckets\main`(for default installation). if git installed before, just clone it to the path.

**Reply by Gitoffthelawn** [reply]
Great info there.  Thank you.

> Yes, for contents of bucket, the following 4 command lines are equivalent...

Initially, is the content also equivalent to the main bucket that comes with a fresh installation of Scoop?

For the next paragraph, let's refer to the main bucket that comes with a fresh installation of Scoop as "A", and the main bucket created with any of the 4 commands you mentioned as "B".

Over time, assuming the user does not invoke one of the commands you mentioned, will that content of "A" and "B" start to diverge because "A" is static, whereas "B" will receive updates?  Or will they both receive updates?

IOW, is there typically a need for the user to invoke one of those 4 commands?

**Reply by HUMORCE** [reply]
> is the content also equivalent to the main bucket that comes with a fresh installation of Scoop?

Sure

> IOW, is there typically a need for the user to invoke one of those 4 commands?

No. After buckets added, Scoop will update them via `git` when appropriate.

> Over time, assuming the user does not invoke one of the commands you mentioned, will that content of "A" and "B" start to diverge because "A" is static, whereas "B" will receive updates? Or will they both receive updates?

- If user does not have a git installation, Scoop will prompt about it.
- If bucket has not been converted to git repo, Scoop will always prompt and try to convert it. (main bucket only)

---

Scoop supports installing apps from the following sources:

- buckets
  - local bucket: `~\scoop\buckets\local`
  - remote bucket
    - official buckets: you can list them via execute `scoop bucket known`
    - 3rd-party buckets: search on github or visit https://scoop.sh/#/buckets
- standalone manifest
  - local file: `scoop install C:\app.json`
  - remote file: `scoop install https://example.org/app.json`, `scoop install https://github.com/ScoopInstaller/Main/raw/refs/heads/master/bucket/7zip.json`

**Reply by Gitoffthelawn** [reply]
> > IOW, is there typically a need for the user to invoke one of those 4 commands?
> 
> No. After buckets added, Scoop will update them via `git` when appropriate.

When you mention *"After buckets added"*, do you mean when Scoop automatically adds the main bucket when Scoop itself is installed?

**Reply by HUMORCE** [reply]
Any bucket that is git repo with remote url set up. you can check added buckets via `scoop bucket list`.

**Reply by Gitoffthelawn** [reply]
> Any bucket that is git repo with remote url set up. you can check added buckets via `scoop bucket list`.

Got it.  Thank you.

And because Scoop automatically converts the default main bucket (generated during Scoop installation) to a git repo when git is installed, the user never even needs to run the command `scoop bucket add main`.  That's a nice touch.
